src/Bicep.Decompiler/TemplateConverter.cs

```diff
@@ -671,34 +670,59 @@ private ObjectSyntax ParseJObject(JObject jObject)
             return SyntaxFactory.CreateObject(properties);
         }
 
-        public ParameterDeclarationSyntax ParseParam(JProperty value)
+        private IEnumerable<SyntaxBase> ProcessDecoratorsWithTransform(IEnumerable<string> functionNames, Func<string, JToken?> valueLookupFunc, Func<(string name, SyntaxBase value), (string name, SyntaxBase value)> transformFunc)
         {
-            var decoratorsAndNewLines = new List<SyntaxBase>();
-
-            // Metadata/description should be first so users see what the parameter is for before
-            // seeing informationi such as a long list of allowed values
-            foreach (var parameterPropertyName in new[] { "metadata", "minValue", "maxValue", "minLength", "maxLength", "allowedValues" })
+            foreach (var functionName in functionNames)
             {
-                if (TryParseJToken(value.Value?[parameterPropertyName]) is SyntaxBase expression)
+                var expressionValue = valueLookupFunc(functionName);
+                if (TryParseJToken(expressionValue) is SyntaxBase expression)
                 {
-                    var functionName = parameterPropertyName == "allowedValues" ? LanguageConstants.ParameterAllowedPropertyName : parameterPropertyName;
+                    var (newFunctionName, newExpression) = transformFunc((functionName, expression));
+
+                    yield return SyntaxFactory.CreateDecorator(newFunctionName, newExpression);
+                    yield return SyntaxFactory.NewlineToken;
+                }
+            }
+        }
 
-                    if (parameterPropertyName == "metadata" &&
+        private IEnumerable<SyntaxBase> ProcessMetadataDecorator(Func<string, JToken?> valueLookupFunc)
+            => ProcessDecoratorsWithTransform(
+                new[] { "metadata" },
+                valueLookupFunc,
+                input => {
+                    var (name, expression) = input;
+
+                    if (name == "metadata" &&
                         expression is ObjectSyntax metadataObject &&
-                        metadataObject.Properties.Any() &&
-                        !metadataObject.Properties.Skip(1).Any() &&
-                        metadataObject.Properties.Single() is ObjectPropertySyntax metadataProperty &&
-                        metadataProperty.TryGetKeyText() == "description")
+                        metadataObject.TryGetPropertyByName("description") is {} descriptionProperty)
```

Comment:
Should we add a check to ensure that the object only contains `description`?

Reply:
Updated - see other comment